packages/@lwc/engine-core/src/framework/api.ts

```diff
@@ -145,42 +152,67 @@ const CustomElementHook: Hooks<VCustomElement> = {
             sel,
             owner: { renderer },
         } = vnode;
-        const elm = renderer.createElement(sel);
-
-        linkNodeToShadow(elm, vnode);
-        createViewModelHook(elm, vnode);
+        const CEBridge = registerTagName(sel, renderer);
+        (CEBridge as any).upgrade = (elm: HTMLElement) => {
+            createViewModelHook(elm, vnode);
+        };
+        const elm = new CEBridge();
```

Comment:
now there is a clear different between custom elements and regular elements, custom elements are newed using the class read from `customElements` registry, while regular elements are created via `document.createElement`. This helps with the SSR as well.

The trick is to provide, right before new it, the upgrade callback that can be executed during the construction path.